File: csharp/DepthFirstTraverser.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;

namespace WikiSearch
{
	public class DepthFirstTraverser
	{
		private WikiPage root;

		public DepthFirstTraverser(WikiPage root)
		{
			this.root = root;
		}

		public List<WikiPage> Traverse()
		{
			List<WikiPage> visited = new List<WikiPage>();
			List<WikiPage> toVisit = new List<WikiPage>();
			toVisit.Add(root);
			while (toVisit.Count > 0) {
				WikiPage node = toVisit[toVisit.Count - 1];
				toVisit.Remove(node);
				if (!visited.Contains(node)) {
					visited.Add(node);
					toVisit.AddRange(node.Children());
				}
			}
			return visited;
		}
	}
}

File: java/src/DepthFirstTraverser.java
import java.util.ArrayList;
import java.util.List;

class DepthFirstTraverser {
	private WikiPage root;

	public DepthFirstTraverser(WikiPage root) {
		this.root = root;
	}

	public List<WikiPage> traverse() {
		List<WikiPage> visited = new ArrayList<WikiPage>();
		List<WikiPage> to_visit = new ArrayList<WikiPage>();
		to_visit.add(root);
		while (to_visit.size() > 0) {
			WikiPage node = to_visit.remove(to_visit.size() - 1);
			if (!visited.contains(node)) {
				visited.add(node);
				to_visit.addAll(node.children());
			}
		}
		return visited;
	}
}
